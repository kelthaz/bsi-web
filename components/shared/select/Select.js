/* eslint-disable complexity */
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styles from './select.module.scss';
import useOnClick from '../../../hooks/useOnClick';
import SvgChevron from '../../svgs/SvgChevron';
// import SvgCross from '../../svgs/SvgCross';

const seleccionaEstilo = (size, inverted) => {
  const finalStyles = [];
  if (inverted) {
    finalStyles.push(styles['arrow-inverted']);
    finalStyles.push(size === 'big' ? styles['select-big-inverted'] : styles['select-small-inverted']);
    finalStyles.push(styles['indicador-inverted']);
    finalStyles.push(styles['indicador-activo-inverted']);
    finalStyles.push(styles['help-text-inverted']);
    finalStyles.push(styles['placeholder-color-inverted']);
  } else {
    finalStyles.push(styles.arrow);
    finalStyles.push(size === 'big' ? styles['select-big'] : styles['select-small']);
    finalStyles.push(styles.indicador);
    finalStyles.push(styles['indicador-activo']);
    finalStyles.push(styles['help-text']);
    finalStyles.push(styles['placeholder-color']);
  }
  return finalStyles;
};

const Select = (props) => {
  const [toggle, setToggle] = useState(false);
  const {
    name,
    formulario,
    size,
    items,
    inverted,
    optional,
    label,
    disabled,
    defaultValue,
    blue,
    showAlwaysErrors,
  } = props;
  const [
    arrrowStyle,
    selectStyle,
    indicadorStyle,
    indicadorActiveStyle,
    helpTextStyle,
    placeholderStyle,
  ] = seleccionaEstilo(size, inverted);
  const { setFieldValue, setFieldTouched, getFieldMeta } = formulario;
  const { error, touched, value } = getFieldMeta(name);

  useEffect(() => {
    if (!value && defaultValue !== null) {
      setFieldValue(name, items[defaultValue]);
    }
  }, []);

  const handleToggle = () => {
    if (!touched && toggle) {
      setFieldTouched(name, true);
    }
    setToggle(!toggle);
  };

  useOnClick(toggle, handleToggle);

  const handleItem = (itemMap) => {
    setFieldValue(name, itemMap);
  };

  return (
    <div className={styles['custom-select']}>
      <button
        type="button"
        className={`svg-button-input-small ${arrrowStyle} ${toggle ? styles['arrow-active'] : ''}`}
        onClick={() => handleToggle()}
        tabIndex="-1"
        disabled={disabled}
      >
        <SvgChevron />
      </button>
      {/* {values[name] && <button
        type="button"
        className={`svg-button-input-small ${styles['deselect-item']}`}
        onClick={() => setFieldValue(name, null)}
      >
        <SvgCross />
      </button>} */}
      <button
        type="button"
        className={`${!toggle && !value ? placeholderStyle : ''} ${blue ? styles['select-small-blue'] : selectStyle} ${
          toggle ? indicadorActiveStyle : indicadorStyle
        } ${touched && error ? styles['indicador-error'] : ''}`}
        onClick={() => !toggle && handleToggle()}
        disabled={disabled}
        tabIndex="0"
        // onFocus={() => !toggle && handleToggle()}
        // onBlur={() => handleToggle()}
      >
        {value ? value.label : label}
      </button>
      <ul role="menu" className={`${styles['select-items']} ${toggle ? '' : styles['select-hide']}`}>
        {items.map((itemMap) => (
          <li role="none" key={itemMap.value}>
            <button
              role="menuitem"
              className={`${styles.item} ${value && value.value === itemMap.value ? styles['item-selected'] : ''}`}
              type="button"
              onClick={() => handleItem(itemMap)}
              tabIndex="-1"
            >
              {itemMap.label}
            </button>
          </li>
        ))}
      </ul>
      <span className={touched && error ? styles['help-text-error'] : helpTextStyle}>
        {touched && error && items.length > 0 ? error : optional && 'Opcional'}
        {showAlwaysErrors && <span>&nbsp;</span>}
      </span>
    </div>
  );
};

Select.propTypes = {
  name: PropTypes.string.isRequired,
  formulario: PropTypes.any.isRequired,
  size: PropTypes.string.isRequired,
  inverted: PropTypes.bool,
  optional: PropTypes.bool,
  items: PropTypes.array.isRequired,
  label: PropTypes.string.isRequired,
  defaultValue: PropTypes.number,
  disabled: PropTypes.bool,
  blue: PropTypes.bool,
  showAlwaysErrors: PropTypes.bool,
};

Select.defaultProps = {
  inverted: false,
  optional: false,
  defaultValue: null,
  disabled: false,
  blue: false,
  showAlwaysErrors: true,
};

export default Select;
