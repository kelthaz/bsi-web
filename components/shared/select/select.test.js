import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';
import Select from './Select';

describe('Pruebas en el componente Select', () => {
  const setValue = jest.fn();
  const setTouched = jest.fn();

  const items = [
    { label: 'Cónyuges', value: 'Cónyuges' },
    { label: 'Concubinos', value: 'oncubinos' },
    { label: 'Hijos', value: 'Hijos' },
    { label: 'Padres', value: 'Padres' },
    { label: 'Suegros', value: 'Suegros' },
    { label: 'Hijos de cónyuge', value: 'Hijos de cónyuge' },
    { label: 'Hermanos', value: 'Hermanos' },
    { label: 'Abuelos', value: 'Abuelos' },
    { label: 'Nietos', value: 'Nietos' },
    { label: 'Cuñados', value: 'Cuñados' },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Debe de mostrarse correctamente el componente', () => {
    // arrange

    const props = {
      name: 'plazo',
      label: 'Seleccione...',
      size: 'small',
      items,
      blue: true,
      format: 'uppercase',
      touched: false,
      value: items[0],
      setValue,
      setTouched,
      inverted: true,
      showAlwaysErrors: false,
    };
    // act
    const wrapper = shallow(<Select {...props} />);
    // assert
    expect(wrapper).toMatchSnapshot();
  });

  it('Debe de llamarse el setValue y el setTouched, con el valor nuevo', async () => {
    // arrange
    const itemEsperado = items[0];
    const props = {
      name: 'plazo',
      label: 'Seleccione...',
      size: 'small',
      items,
      blue: true,
      format: 'uppercase',
      touched: false,
      value: null,
      setValue,
      setTouched,
      inverted: false,
    };
    const wrapper = shallow(<Select {...props} />);
    // act
    await act(async () => {
      wrapper.find('button').at(0).simulate('click');
    });
    await act(async () => {
      wrapper
        .find('ul')
        .find('button')
        .at(0)
        .simulate('click', { ...itemEsperado });
    });

    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(itemEsperado);
    expect(setTouched).toHaveBeenCalledTimes(1);
    expect(setTouched).toHaveBeenCalledWith(true);
  });
  it('Debe de llamarse el setValue y el setTouched 1 vez, y por último cerrar el select ', async () => {
    // arrange
    const segundoItem = items[1];
    const props = {
      name: 'plazo',
      label: 'Seleccione...',
      size: 'big',
      items,
      format: 'uppercase',
      touched: false,
      value: null,
      setValue,
      setTouched,
      inverted: false,
    };
    const wrapper = shallow(<Select {...props} />);
    // act
    await act(async () => {
      wrapper.find('button').at(0).simulate('click');
    });

    await act(async () => {
      wrapper.find('button').at(0).simulate('click');
    });
    await act(async () => {
      wrapper
        .find('ul')
        .find('button')
        .at(1)
        .simulate('click', { ...segundoItem });
    });
    await act(async () => {
      wrapper.find('button').at(1).simulate('click');
    });

    // assert
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(segundoItem);
    expect(setTouched).toHaveBeenCalledTimes(2);
  });
  it('Debe de no mostrarse el error debido a que no tiene items ', async () => {
    // arrange
    const error = 'Seleccione una opción';
    const props = {
      name: 'plazo',
      label: 'Seleccione...',
      size: 'big',
      error,
      items: [],
      format: 'uppercase',
      touched: false,
      value: null,
      setValue,
      setTouched,
      inverted: true,
    };
    const wrapper = shallow(<Select {...props} />);
    // act
    await act(async () => {
      wrapper.find('button').at(1).simulate('click');
    });
    const span = wrapper.find('span').at(0).text().trim();
    // assert
    expect(setValue).toHaveBeenCalledTimes(0);
    expect(setTouched).toHaveBeenCalledTimes(0);
    expect(span).toBe('');
  });

  it('Debe de no mostrarse el error debido a que esta deshabilitado', async () => {
    // arrange
    const props = {
      name: 'plazo',
      label: 'Seleccione...',
      size: 'big',
      items: [],
      format: 'uppercase',
      touched: false,
      value: null,
      setValue,
      setTouched,
      inverted: true,
      disabled: true,
    };
    const wrapper = shallow(<Select {...props} />);
    // act
    const span = wrapper.find('span').at(0).text().trim();
    // assert
    expect(setValue).toHaveBeenCalledTimes(0);
    expect(setTouched).toHaveBeenCalledTimes(0);
    expect(span).toBe('');
  });
  it('Debe de mostrarse el error, debido a que esta habilitado para eso', async () => {
    // arrange
    const errorEsperado = 'Seleccione una opción';
    const props = {
      name: 'plazo',
      label: 'Seleccione...',
      size: 'big',
      items,
      error: errorEsperado,
      format: 'uppercase',
      touched: true,
      value: null,
      setValue,
      setTouched,
      inverted: true,
      disabled: false,
    };
    const wrapper = shallow(<Select {...props} />);
    // act
    const span = wrapper.find('span').at(0).text().trim();
    // assert
    expect(span).toBe(errorEsperado);
  });
});
