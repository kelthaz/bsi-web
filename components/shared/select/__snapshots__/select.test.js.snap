// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente Select Debe de mostrarse correctamente el componente 1`] = `
<div
  className="custom-select"
>
  <button
    className="svg-button-input-small arrow-inverted "
    disabled={false}
    onClick={[Function]}
    tabIndex="-1"
    type="button"
  >
    <SvgChevron />
  </button>
  <button
    className=" select-small-blue indicador-inverted "
    disabled={false}
    name="plazo"
    onClick={[Function]}
    tabIndex="0"
    type="button"
  >
    Cónyuges
  </button>
  <ul
    className="select-items select-hide"
    role="menu"
  >
    <li
      key="Cónyuges"
      role="none"
    >
      <button
        className="item item-selected"
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Cónyuges
      </button>
    </li>
    <li
      key="oncubinos"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Concubinos
      </button>
    </li>
    <li
      key="Hijos"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Hijos
      </button>
    </li>
    <li
      key="Padres"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Padres
      </button>
    </li>
    <li
      key="Suegros"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Suegros
      </button>
    </li>
    <li
      key="Hijos de cónyuge"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Hijos de cónyuge
      </button>
    </li>
    <li
      key="Hermanos"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Hermanos
      </button>
    </li>
    <li
      key="Abuelos"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Abuelos
      </button>
    </li>
    <li
      key="Nietos"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Nietos
      </button>
    </li>
    <li
      key="Cuñados"
      role="none"
    >
      <button
        className="item "
        onClick={[Function]}
        role="menuitem"
        tabIndex="-1"
        type="button"
      >
        Cuñados
      </button>
    </li>
  </ul>
  <span
    className="help-text-error"
  />
</div>
`;
